Adding support for `$gem-number` template for sending gems

shuaibq3 opened the pull request:
Closes #29 

# Description
In current implementation, gem count is determined by the `:gem:` string in the discord message. It introduces the template `gems-number` to provide gems with a number, rather than having to type the string `:gem:` repeatedly. example: `gems-3 for the help` will return the gem count as 3 from this message.

This template works besides the existing implementation. It adds the template as an addition, without changing the existing implementation. This means, a user can use both the current way, and the `template` way to send gems.

Unit tests have been added for counting gems from message string. The tests are divided into 3 test classes (in the same file), validating current implementation behaviour, validating template string behaviour, and validating using both in the same message. 

Giving a number of examples below to clarify the implemented behaviour:
- `gems-3 for the help` will return 3 gem count.
- `gems-35 for the help` will return 35 gem count.
- `gems-3 gems-35 for the help` will return 3 gem count.
- `gems-3gems-35 for the help` will return 0 gem count.
- `gems-----35 for the help` will return 0 gem count.
- `gems-10.8 for the help` will return 0 gem count.
- `gems-abcd gems-35 for the help` will return 35 gem count.
- Using both `:gem:` and `gems-number` returns the count of `:gem:` and disregards the template.

Please check the tests to understand the implemented behaviour for these cases.

sakibstark11 commented on src/command.py:
```diff
@@ -49,6 +49,43 @@ def __repr__(self):
                f"Gem message: {self.gem_message}"
 
 
+class GemsCounterFromMessage:
```
Do we really need a whole class for this?

shuaibq3 replied:
i had the same question. decided to create a class considering 2 things:
- making a class gives more manageability in terms of adding more templates to give gems. we can do that if we want.
- why not, making a class does harm in anyway.
please let me know if you have other thoughts in mind. i only shared my thoughts while coding, may be missing something?

sakibstark11 replied:
Semantically speaking a class makes sense when you have locallised attributes that belong to the instance of that class, in this case we are not encapsulating/locallising anything, infact, what would be better if we create a separate module/file in python called gem-message-template-parser.py which can have an entry function that gets passed to the gemMessage class, meaning in the future, we can switch that out for a different gem message parser. Move those tiny functions into a file, create a dependency injection system for the gemmessage whatever class and it might look really clean :D

shuaibq3 replied:
based on discussion, moved the message parsing implementation to a separate file in commit [ef68eb3](https://github.com/CraftsmenLtd/DiscordGems/pull/30/commits/ef68eb35c6dbc07eab17f1ab69a754f1788c3240)

sakibstark11 commented on src/command.py:
```diff
@@ -49,6 +49,43 @@ def __repr__(self):
+        if gem_count_string.isdigit():
+            return int(gem_count_string)
+        try:
+            gem_count_float = float(gem_count_string)
```
To me this feels like it should just fail instead of trying to convert to float then int?
```python
try:
    return int(gem_count_string)
except some_error:
    return 0
```

shuaibq3 replied:
yes, i had considered if we return 0 instead of int value of the float. according to the latest push, it returns 0. test modified accordingly.

sakibstark11 commented on src/command.py:
```diff
@@ -49,6 +49,43 @@ def __repr__(self):
+        if message.count(gem_message_string_anchor) == 0:
+            return 0
+
+        gem_message_words = message.split()
```
Could this not be done with regex?
We are iterating multiple times, discord does have a hard limit on the response time of its hooks. This Probably leads me to suggest if we should separate the gem option from the message option?

shuaibq3 replied:
that is an awesome idea :D implemented regex in commit [682d44d](https://github.com/CraftsmenLtd/DiscordGems/pull/30/commits/682d44df0c99efc675bebf8e3033d58f82061f4a) and it seems to have reduced the lines of code significantly.

i did consider regex at first, but was not aware that we can make a list of matching strings with regex. i was mistaken to think after matching with regex, i will have to use split and match string anyway to parse and get the number. of course that was a mistake, and learned something new :D

thank you mate for the great feedback :+1:

sakibstark11 commented:
Firstly thank you for raising this PR. It's much appreciated. A couple things
With your approach `$gem-` must be used to submit gems. I like this idea because it's the user voluntarily and consciously deciding on the number of gems. I am not a fan of the dollar sign though, it feels unintuitive to me. Do you have strong feelings for it?  Can we talk about it?

Secondly, thank you for writing some tests for this change. <3

Thirdly, `GemsCounterFromMessage`; do we really need an entire class for this logic? given its not encapsulating any information/data/attribute within the class. Left comments Individually.

Again thank you.

shuaibq3 commented:
> Firstly thank you for raising this PR. It's much appreciated. A couple things With your approach `$gem-` must be used to submit gems. I like this idea because it's the user voluntarily and consciously deciding on the number of gems. I am not a fan of the dollar sign though, it feels unintuitive to me. Do you have strong feelings for it? Can we talk about it?
> 
> Secondly, thank you for writing some tests for this change. <3
> 
> Thirdly, `GemsCounterFromMessage`; do we really need an entire class for this logic? given its not encapsulating any information/data/attribute within the class. Left comments Individually.
> 
> Again thank you.

regarding the use of `$`, i am not a big fan of it either. the fact that you have to press a `shift` and move the finger to the numbers row of the keyboard is so annoying when in a smooth flow of typing. but that is precisely why i added this, so that the user is aware of what they are typing when they are giving gems. or at least increases the awareness when typing.

i think we can remove this if you suggest, as with a wrong gem input the user can send another message with gem adjustments one anyway, unless of course, the previous message consumed all the gems for the day :P but that isnt a big deal either. please let me know how you would suggest we proceed with this :)

regarding making a class, added a [comment](https://github.com/CraftsmenLtd/DiscordGems/pull/30#discussion_r1299362385) in the review section.

once more, thank you so much for the valuable feedbacks. and please do provide more if you have :D

shuaibq3 commented:
again, thank you for the valuable feedbacks! according to discussion, changed the template to `number-gems` from `$gem-number` in commit [b2b9e35](https://github.com/CraftsmenLtd/DiscordGems/pull/30/commits/b2b9e35f0a757e1d77f991c6cec3336aec843fee). changed the PR description to reflect the latest changes

sakibstark11 commented on src/message_gems_parser.py:
```diff
@@ -0,0 +1,18 @@
+import re
+
+def get_gem_count_from_gem_string(message: str):
+    gem_template_string_anchor = '-gems'
```
i thought we agreed je gems- ta e better?

shuaibq3 replied:
yes i misunderstood.. fixed in commit [13ca5bf](https://github.com/CraftsmenLtd/DiscordGems/pull/30/commits/13ca5bfb315f0b06ca2874c32a61aa1f32382005)
